File: crates/sidecar-kit/tests/mock_sidecar.py
"""Mock sidecar for sidecar-kit lifecycle tests.

Speaks the JSONL Frame protocol (tag = "t").

Modes:
  default         - hello → run → event → final
  large_stream    - hello → run → 100 events → final
  error_midstream - hello → run → event → fatal
  empty_work_order- hello → run → final (no events)
  tool_call       - hello → run → tool_call event → tool_result event → final
  multi_run       - hello → (run → event → final) repeated for each run on stdin
  slow            - hello → run → event (0.5s delay each) → final
  crash           - hello → run → event → exit(1)
"""
import sys
import json
import time

mode = sys.argv[1] if len(sys.argv) > 1 else "default"


def make_hello():
    return {
        "t": "hello",
        "contract_version": "abp/v0.1",
        "backend": {"id": "mock-lifecycle", "version": "0.1"},
        "capabilities": {"streaming": True},
    }


def read_run():
    line = sys.stdin.readline()
    if not line:
        sys.exit(0)
    run = json.loads(line)
    return run["id"], run.get("work_order", {})


def make_event(ref_id, payload):
    return {"t": "event", "ref_id": ref_id, "event": payload}


def make_final(ref_id):
    return {
        "t": "final",
        "ref_id": ref_id,
        "receipt": {"status": "complete", "ref_id": ref_id},
    }


def emit(obj):
    print(json.dumps(obj), flush=True)


# ---- modes ----------------------------------------------------------------

if mode == "default":
    emit(make_hello())
    ref_id, wo = read_run()
    emit(make_event(ref_id, {"type": "progress", "step": 1}))
    emit(make_event(ref_id, {"type": "progress", "step": 2}))
    emit(make_final(ref_id))

elif mode == "large_stream":
    emit(make_hello())
    ref_id, wo = read_run()
    for i in range(100):
        emit(make_event(ref_id, {"type": "progress", "index": i}))
    emit(make_final(ref_id))

elif mode == "error_midstream":
    emit(make_hello())
    ref_id, wo = read_run()
    emit(make_event(ref_id, {"type": "progress", "step": 1}))
    emit({"t": "fatal", "ref_id": ref_id, "error": "processing failed"})

elif mode == "empty_work_order":
    emit(make_hello())
    ref_id, wo = read_run()
    emit(make_final(ref_id))

elif mode == "tool_call":
    emit(make_hello())
    ref_id, wo = read_run()
    emit(make_event(ref_id, {"type": "tool_call", "tool": "read_file", "args": {"path": "test.txt"}}))
    emit(make_event(ref_id, {"type": "tool_result", "tool": "read_file", "result": "file contents"}))
    emit(make_final(ref_id))

elif mode == "multi_run":
    emit(make_hello())
    # Handle up to 3 sequential runs
    for _ in range(3):
        try:
            line = sys.stdin.readline()
            if not line:
                break
            run = json.loads(line)
            ref_id = run["id"]
            emit(make_event(ref_id, {"type": "progress", "step": 1}))
            emit(make_final(ref_id))
        except Exception:
            break

elif mode == "slow":
    emit(make_hello())
    ref_id, wo = read_run()
    emit(make_event(ref_id, {"type": "progress", "step": "start"}))
    time.sleep(0.3)
    emit(make_event(ref_id, {"type": "progress", "step": "middle"}))
    time.sleep(0.3)
    emit(make_event(ref_id, {"type": "progress", "step": "end"}))
    emit(make_final(ref_id))

elif mode == "crash":
    emit(make_hello())
    ref_id, wo = read_run()
    emit(make_event(ref_id, {"type": "progress", "step": 1}))
    sys.stdout.flush()
    import os
    os._exit(1)

else:
    print(f"Unknown mode: {mode}", file=sys.stderr)
    sys.exit(1)

File: crates/sidecar-kit/tests/bin/mock_sidecar.rs
// SPDX-License-Identifier: MIT OR Apache-2.0
//! Rust mock sidecar for sidecar-kit lifecycle tests.

use serde_json::{Value, json};
use std::env;
use std::io::{self, BufRead, Write};
use std::thread;
use std::time::Duration;

fn make_hello() -> Value {
    json!({
        "t": "hello",
        "contract_version": "abp/v0.1",
        "backend": {"id": "mock-lifecycle", "version": "0.1"},
        "capabilities": {"streaming": true}
    })
}

fn read_run() -> Option<(String, Value)> {
    let mut line = String::new();
    if io::stdin().lock().read_line(&mut line).ok()? == 0 {
        return None;
    }
    let run: Value = serde_json::from_str(&line).ok()?;
    Some((
        run["id"].as_str()?.to_owned(),
        run.get("work_order").cloned().unwrap_or_else(|| json!({})),
    ))
}

fn make_event(ref_id: &str, payload: Value) -> Value {
    json!({"t": "event", "ref_id": ref_id, "event": payload})
}

fn make_final(ref_id: &str) -> Value {
    json!({"t": "final", "ref_id": ref_id, "receipt": {"status": "complete", "ref_id": ref_id}})
}

fn emit(value: Value) {
    println!(
        "{}",
        serde_json::to_string(&value).expect("serialize frame")
    );
    io::stdout().flush().expect("flush stdout");
}

fn sleep_ms(ms: u64) {
    thread::sleep(Duration::from_millis(ms));
}

fn main() {
    let mode = env::args().nth(1).unwrap_or_else(|| "default".to_owned());

    match mode.as_str() {
        "default" => {
            emit(make_hello());
            let Some((ref_id, _)) = read_run() else {
                return;
            };
            emit(make_event(&ref_id, json!({"type": "progress", "step": 1})));
            emit(make_event(&ref_id, json!({"type": "progress", "step": 2})));
            emit(make_final(&ref_id));
        }
        "large_stream" => {
            emit(make_hello());
            let Some((ref_id, _)) = read_run() else {
                return;
            };
            for i in 0..100 {
                emit(make_event(&ref_id, json!({"type": "progress", "index": i})));
            }
            emit(make_final(&ref_id));
        }
        "error_midstream" => {
            emit(make_hello());
            let Some((ref_id, _)) = read_run() else {
                return;
            };
            emit(make_event(&ref_id, json!({"type": "progress", "step": 1})));
            emit(json!({"t": "fatal", "ref_id": ref_id, "error": "processing failed"}));
        }
        "empty_work_order" => {
            emit(make_hello());
            let Some((ref_id, _)) = read_run() else {
                return;
            };
            emit(make_final(&ref_id));
        }
        "tool_call" => {
            emit(make_hello());
            let Some((ref_id, _)) = read_run() else {
                return;
            };
            emit(make_event(
                &ref_id,
                json!({"type": "tool_call", "tool": "read_file", "args": {"path": "test.txt"}}),
            ));
            emit(make_event(
                &ref_id,
                json!({"type": "tool_result", "tool": "read_file", "result": "file contents"}),
            ));
            emit(make_final(&ref_id));
        }
        "multi_run" => {
            emit(make_hello());
            for _ in 0..3 {
                let Some((ref_id, _)) = read_run() else {
                    break;
                };
                emit(make_event(&ref_id, json!({"type": "progress", "step": 1})));
                emit(make_final(&ref_id));
            }
        }
        "slow" => {
            emit(make_hello());
            let Some((ref_id, _)) = read_run() else {
                return;
            };
            emit(make_event(
                &ref_id,
                json!({"type": "progress", "step": "start"}),
            ));
            sleep_ms(300);
            emit(make_event(
                &ref_id,
                json!({"type": "progress", "step": "middle"}),
            ));
            sleep_ms(300);
            emit(make_event(
                &ref_id,
                json!({"type": "progress", "step": "end"}),
            ));
            emit(make_final(&ref_id));
        }
        "crash" => {
            emit(make_hello());
            let Some((ref_id, _)) = read_run() else {
                return;
            };
            emit(make_event(&ref_id, json!({"type": "progress", "step": 1})));
            std::process::exit(1);
        }
        _ => {
            eprintln!("Unknown mode: {mode}");
            std::process::exit(1);
        }
    }
}
